datastore/nipost/nipost.go

```diff
@@ -0,0 +1,125 @@
+package nipost
+
+import (
+	"fmt"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/datastore"
+	"github.com/spacemeshos/go-spacemesh/sql"
+)
+
+func AddChallenge(db *datastore.LocalDB, nodeID types.NodeID, ch *types.NIPostChallenge) error {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, nodeID.Bytes())
+		stmt.BindInt64(2, int64(ch.PublishEpoch))
+		stmt.BindInt64(3, int64(ch.Sequence))
+		stmt.BindBytes(4, ch.PrevATXID.Bytes())
+		stmt.BindBytes(5, ch.PositioningATX.Bytes())
+		if ch.CommitmentATX != nil {
+			stmt.BindBytes(6, ch.CommitmentATX.Bytes())
+		} else {
+			stmt.BindNull(6)
+		}
+		if ch.InitialPost != nil {
+			stmt.BindInt64(7, int64(ch.InitialPost.Nonce))
+			stmt.BindBytes(8, ch.InitialPost.Indices)
+			stmt.BindInt64(9, int64(ch.InitialPost.Pow))
+		} else {
+			stmt.BindInt64(7, -1)
+			stmt.BindNull(8)
+			stmt.BindInt64(9, -1)
+		}
+	}
+
+	if _, err := db.Exec(`
+		insert into nipost (id, epoch, sequence, prev_atx, pos_atx, commit_atx, post_nonce, post_indices, post_pow)
+		values (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9);`, enc, nil,
+	); err != nil {
+		return fmt.Errorf("insert nipost challenge for %s pub-epoch %d: %w", nodeID, ch.PublishEpoch, err)
+	}
+	return nil
+}
+
+func UpdateChallenge(db *datastore.LocalDB, nodeID types.NodeID, ch *types.NIPostChallenge) error {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, nodeID.Bytes())
+		stmt.BindInt64(2, int64(ch.PublishEpoch))
+		stmt.BindInt64(3, int64(ch.Sequence))
+		stmt.BindBytes(4, ch.PrevATXID.Bytes())
+		stmt.BindBytes(5, ch.PositioningATX.Bytes())
+		if ch.CommitmentATX != nil {
+			stmt.BindBytes(6, ch.CommitmentATX.Bytes())
+		} else {
+			stmt.BindNull(6)
+		}
+		if ch.InitialPost != nil {
+			stmt.BindInt64(7, int64(ch.InitialPost.Nonce))
+			stmt.BindBytes(8, ch.InitialPost.Indices)
+			stmt.BindInt64(9, int64(ch.InitialPost.Pow))
+		} else {
+			stmt.BindInt64(7, -1)
+			stmt.BindNull(8)
+			stmt.BindInt64(9, -1)
+		}
+	}
```

Comment:
I am not convinced to extract the encoding function from `Add` and `Update`. The are only the same by coincidence and do not depend on `nodeID` and `types.NIPostChallenge` but rather on the specific SQL query of `Add` and `Update`.

If the encoding is extracted and shared between both methods changing one of the two SQL queries might break the encoding function for it or the other 😕

Reply:
I don't think it's a coincidence. The whole point of extracting is to have them equal to avoid bugs :shrug:. It can always be unsplit if there is actually a need for different implementations.

Reply:
The sql query could state the columns in any order. The reason the code is the same is because I assigned the same columns to the same indices, but I didn't have to.

Reply:
I extracted the encoding function for now, but I might split it again in the next PR.